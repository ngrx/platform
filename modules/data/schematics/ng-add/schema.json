{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxData",
  "title": "NgRx Data Schema",
  "type": "object",
  "properties": {
    "effect": {
      "type": "boolean",
      "default": true,
      "description":
        "When false use the EntityDataModuleWithoutEffects module instead of EntityDataModule."
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/data as dependency to package.json (e.g., --skipPackageJson)."
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to the module.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "module": {
      "type": "string",
      "default": "app",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "migrateNgrxData": {
      "type": "boolean",
      "default": false,
      "description": "Migrate from ngrx-data, will rename modules.",
      "alias": "migrate"
    },
    "entityConfig": {
      "type": "boolean",
      "default": true,
      "description": "Create the Entity config file"
    }
  },
  "required": []
}
