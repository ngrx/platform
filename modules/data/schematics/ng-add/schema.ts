export interface Schema {
  path?: string;
  effects?: boolean;
  skipPackageJson?: boolean;
  project?: string;
  module?: string;
  migrateNgrxData?: boolean;
  entityConfig?: boolean;
}
