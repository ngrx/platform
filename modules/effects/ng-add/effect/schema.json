{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEffect",
  "title": "NgRx Effect Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the effect.",
      "type": "string"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "module": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "root": {
      "type": "boolean",
      "default": false,
      "description":
        "Flag to indicate whether the effects are registered as root."
    },
    "feature": {
      "type": "boolean",
      "default": false,
      "description": "Flag to indicate if part of a feature schematic."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group effects file within 'effects' folder",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
