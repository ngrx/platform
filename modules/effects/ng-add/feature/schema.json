{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxFeature",
  "title": "NgRx Feature State Options Schema",
  "type": "object",
  "properties": {
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "name": {
      "description": "The name of the feature.",
      "type": "string"
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "module": {
      "type": "string",
      "description": "Specifies the declaring module.",
      "aliases": ["m"]
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "reducers": {
      "type": "string",
      "description": "Specifies the reducers file.",
      "aliases": ["r"]
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description":
        "Group actions, reducers and effects within relative subfolders",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
