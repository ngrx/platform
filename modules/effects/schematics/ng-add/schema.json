{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxRootEffect",
  "title": "NgRx Root Effect Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the effect.",
      "type": "string",
      "default": "App"
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description":
        "Do not add @ngrx/effects as dependency to package.json (e.g., --skipPackageJson)."
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the effect.",
      "visible": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "spec": {
      "type": "boolean",
      "default": true,
      "description": "Specifies if a spec file is generated."
    },
    "module": {
      "type": "string",
      "default": "app",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath"
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group effects file within 'effects' folder",
      "aliases": ["g"]
    }
  },
  "required": []
}
