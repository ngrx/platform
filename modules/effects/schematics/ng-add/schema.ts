export interface Schema {
  name: string;
  skipPackageJson?: boolean;
  path?: string;
  project?: string;
  flat?: boolean;
  spec?: boolean;
  module?: string;
  group?: boolean;
}
