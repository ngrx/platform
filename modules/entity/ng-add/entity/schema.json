{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEntity",
  "title": "NgRx Entity Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the entity.",
      "type": "string"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "spec": {
      "type": "boolean",
      "description": "Specifies if a spec file is generated.",
      "default": true
    },
    "reducers": {
      "type": "string",
      "description": "Specifies the reducers file.",
      "aliases": ["r"]
    },
    "module": {
      "type": "string",
      "description": "Specifies the declaring module.",
      "aliases": ["m"]
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description":
        "Group actions, reducers and effects within relative subfolders",
      "aliases": ["g"]
    }
  },
  "required": ["name"]
}
