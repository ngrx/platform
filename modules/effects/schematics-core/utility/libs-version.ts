export const platformVersion = '^15.0.0-rc.0';
