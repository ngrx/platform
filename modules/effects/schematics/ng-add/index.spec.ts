import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import { getFileContent } from '@schematics/angular/utility/test';
import * as path from 'path';
import { Schema as RootEffectOptions } from './schema';
import {
  getTestProjectPath,
  createWorkspace,
  createAppModuleWithEffects,
} from '../../../schematics-core/testing';

describe('Effect ng-add Schematic', () => {
  const schematicRunner = new SchematicTestRunner(
    '@ngrx/effects',
    path.join(__dirname, '../collection.json')
  );

  const defaultOptions: RootEffectOptions = {
    name: 'foo',
    skipPackageJson: false,
    project: 'bar',
    spec: true,
    module: 'app',
    flat: false,
    group: false,
  };

  const projectPath = getTestProjectPath();

  let appTree: UnitTestTree;

  beforeEach(() => {
    appTree = createWorkspace(schematicRunner, appTree);
  });

  it('should update package.json', () => {
    const options = { ...defaultOptions };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const packageJson = JSON.parse(tree.readContent('/package.json'));

    expect(packageJson.dependencies['@ngrx/effects']).toBeDefined();
  });

  it('should skip package.json update', () => {
    const options = { ...defaultOptions, skipPackageJson: true };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const packageJson = JSON.parse(tree.readContent('/package.json'));

    expect(packageJson.dependencies['@ngrx/effects']).toBeUndefined();
  });

  it('should create an effect', () => {
    const options = { ...defaultOptions };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const files = tree.files;
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.effects.spec.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.effects.ts`)
    ).toBeGreaterThanOrEqual(0);
  });

  it('should be provided by default', () => {
    const options = { ...defaultOptions };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/import { FooEffects } from '.\/foo\/foo.effects'/);
  });

  it('should import into a specified module', () => {
    const options = { ...defaultOptions, module: 'app.module.ts' };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/import { FooEffects } from '.\/foo\/foo.effects'/);
  });

  it('should fail if specified module does not exist', () => {
    const options = {
      ...defaultOptions,
      module: `${projectPath}/src/app/app.moduleXXX.ts`,
    };
    let thrownError: Error | null = null;
    try {
      schematicRunner.runSchematic('effects', options, appTree);
    } catch (err) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should respect the spec flag', () => {
    const options = { ...defaultOptions, spec: false };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const files = tree.files;
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.effects.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.effects.spec.ts`)
    ).toEqual(-1);
  });

  it('should register the root effect in the provided module', () => {
    const options = { ...defaultOptions };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);

    expect(content).toMatch(/EffectsModule\.forRoot\(\[FooEffects\]\)/);
  });

  it('should add an effect to the empty array of registered effects', () => {
    const storeModule = `${projectPath}/src/app/store.module.ts`;
    const options = {
      ...defaultOptions,
      module: 'store.module.ts',
    };
    appTree = createAppModuleWithEffects(
      appTree,
      storeModule,
      'EffectsModule.forRoot([])'
    );

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const content = tree.readContent(storeModule);

    expect(content).toMatch(/EffectsModule\.forRoot\(\[FooEffects\]\)/);
  });

  it('should add an effect to the existing registered root effects', () => {
    const storeModule = `${projectPath}/src/app/store.module.ts`;
    const options = {
      ...defaultOptions,
      module: 'store.module.ts',
    };
    appTree = createAppModuleWithEffects(
      appTree,
      storeModule,
      'EffectsModule.forRoot([UserEffects])'
    );

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const content = tree.readContent(storeModule);

    expect(content).toMatch(
      /EffectsModule\.forRoot\(\[UserEffects, FooEffects\]\)/
    );
  });

  it('should not add an effect to registered effects defined with a variable', () => {
    const storeModule = `${projectPath}/src/app/store.module.ts`;
    const options = { ...defaultOptions, module: 'store.module.ts' };
    appTree = createAppModuleWithEffects(
      appTree,
      storeModule,
      'EffectsModule.forRoot(effects)'
    );

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const content = tree.readContent(storeModule);

    expect(content).not.toMatch(/EffectsModule\.forRoot\(\[FooEffects\]\)/);
  });

  it('should group within an "effects" folder if group is set', () => {
    const options = { ...defaultOptions, flat: true, spec: false, group: true };

    const tree = schematicRunner.runSchematic('ng-add', options, appTree);
    const files = tree.files;
    expect(
      files.indexOf(`${projectPath}/src/app/effects/foo.effects.ts`)
    ).toBeGreaterThanOrEqual(0);
  });
});
