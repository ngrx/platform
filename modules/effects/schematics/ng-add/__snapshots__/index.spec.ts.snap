// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Effects ng-add Schematic Effects ng-add Schematic for standalone application provides full effects setup 1`] = `
"import { ApplicationConfig, provideZoneChangeDetection } from '@angular/core';
import { provideEffects } from '@ngrx/effects';
import { FooEffects } from './foo/foo.effects';

export const appConfig: ApplicationConfig = {
  providers: [provideZoneChangeDetection({ eventCoalescing: true }), provideEffects(FooEffects)]
};
"
`;

exports[`Effects ng-add Schematic Effects ng-add Schematic for standalone application provides minimal effects setup 1`] = `
"import { ApplicationConfig, provideZoneChangeDetection } from '@angular/core';
import { provideEffects } from '@ngrx/effects';

export const appConfig: ApplicationConfig = {
  providers: [provideZoneChangeDetection({ eventCoalescing: true }), provideEffects()]
};
"
`;
